File: src/test/java/com/lsd/diagram/GroupByNewPageEventsTest.java
package com.lsd.diagram;

import com.lsd.events.Message;
import com.lsd.events.Newpage;
import com.lsd.events.PageTitle;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThat;

class GroupByNewPageEventsTest {

    @Test
    void splitsStreamByNewPageEvents() {
        var message1 = Message.builder().id("1").from("A").to("B").build();
        var message2 = Message.builder().id("2").from("B").to("A").build();
        var message3 = Message.builder().id("3").from("B").to("C").build();
        var newpage = Newpage.titled("Part B");

        var sequenceEvents = List.of(message1, message2, newpage, message3);

        var splitEvents = sequenceEvents.stream().collect(new GroupByNewPageEvents());

        assertThat(splitEvents).containsExactly(
                List.of(message1, message2), 
                List.of(PageTitle.titled("Part B"), message3)
        );
    }

}

File: src/test/kotlin/com/lsd/diagram/GroupByNewPageEventsTest.kt
package com.lsd.diagram

import com.lsd.events.*
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

internal class GroupByNewPageEventsTest {
    
    @Test
    fun splitsStreamByNewPageEvents() {
        val message1 = Message(id = "1", label = "", from = "B", to = "A")
        val message2 = Message(id = "2", label = "", from = "A", to = "B")
        val message3 = Message(id = "3", label = "", from = "B", to = "C")

        val sequenceEvents = listOf(
            message1,
            message2,
            Newpage.titled("Part B"),
            message3
        )

        assertThat(sequenceEvents.groupedByPages()).containsExactly(
            mutableListOf(
                message1, 
                message2
            ),
            mutableListOf(
                PageTitle.titled("Part B"),
                message3)
        )
    }
}